src/redistributors/morpho/Distributor.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity 0.8.23;
+
+import { ERC20Helper } from "../../../lib/erc20-helper/src/ERC20Helper.sol";
+
+import { IMorphoVaultFactoryLike, IMorphoVaultLike, IWrappedMLike } from "./Dependencies.sol";
+
+import { AccountFundsDistribution } from "./AccountFundsDistribution.sol";
+import { VaultFundsDistribution } from "./VaultFundsDistribution.sol";
+
+/**
+ * @title Morpho Distributor (Dual-Cascade, Cumulative Distribution)
+ * @dev   A contract for distributing Morpho yield across accounts within vaults.
+ * @dev   Each distribution cascade is an ERC2222 model relying on tracking per-account "corrections".
+ */
+contract Distributor is AccountFundsDistribution, VaultFundsDistribution {
+    error InvalidVault();
+    error TransferFailed();
+    error TransferFromFailed();
+
+    address public immutable wrappedMToken;
+    address public immutable morphoBlue;
+    address public immutable morphoVaultFactory;
+
+    uint256 internal _lastWrappedMBalance;
+
+    mapping(address vault => uint256 lastCumulativeDistribution) internal _lastCumulativeDistributions;
+    mapping(address vault => mapping(address account => uint256 claimed)) internal _claims;
+
+    constructor(address wrappedMToken_, address morphoBlue_, address morphoVaultFactory_) {
+        wrappedMToken = wrappedMToken_;
+        morphoBlue = morphoBlue_;
+        morphoVaultFactory = morphoVaultFactory_;
+    }
+
+    /// @notice Allow `msg.sender` to deposit an amount of Morpho Vault shares to get access to WrappedM yield.
+    function deposit(address vault_, uint256 vaultShares_) external {
+        _revertIfNotMorphoVault(vault_);
```

Comment:
additionally vault needs to specifically wM asset vault, otherwise Distributor is susceptible to attack. 

Additionally, vaults can be whitelisted by the distributor 'curator'

Reply:
The next PR solves for this since it computes the WM liquidity of a vault to rebalance, and such a vault would have 0 wM as assets in the Morpho Blue contract, and it's shares would we worthless in the Distributor. In nay case, I will add it to this PR and then remove iot in the next, for completeness.